src/react/11-state.md

```diff
@@ -0,0 +1,270 @@
+
+```
+const [ viewModel, setViewModel ] = useState(
+    {
+        showModal: false,
+        currentUser: {
+            name: 'Mike Myers'
+            id: 42
+        },
+        highlightItems: [1, 12, 15, 34, 66]
+    }
+)
+```
+
+**Important:** When using objects or arrays, you must provide a new merged state when using the setter. `useState` is immutable and will always replace any internal state.
+
+*For class based components see [this.setState()](https://reactjs.org/docs/state-and-lifecycle.html)*
+
+## Scope of State ##
+
+`useState` is scoped to the component it is declared in. For example, this simple component stores a simple string `displayText` and diplays its value as the user makes changes to a text input.
+
+```
+import React, { useState } from 'react'
+
+function DisplayComponent() {
+    const [ displayText, setDisplayText ] = useState('')
+
+    function onSetText(event) {
+        setDisplayText(event.target.value)
+    }
+
+    return (
+        <div>
+            <h1>You entered: {displayText}</h1>
+            <input type="text" value={displayText} onChange={onSetText}/>
+        </div>
+    )
+}
+```
+
+Now imagine our layout requires a button to confirm the `displayText` change, but the button appears in a different part of the page! How will we modify the state in a different component?
+
+Lucky for us, values and setters from `useState` are standard javascript variables and can be passed as `props`. We can create a 'Container' component to hold the state where both Display and Button components have access. *Consider the following structure:*
+
+```
+function WrapperContainer() {
+    //TODO: setup the state variables
+    return (
+        <div>
+            <div className="layout-left">
+                <DisplayComponent/>
+            </div>
+            <div className="layout-right">
+                <ButtonComponent/>
+            </div>
+        </div>
+    )
+}
+
+function DisplayComponent(props) {
+    return (
+        <div>
+            <h1>You entered:</h1>
+            <input type="text"/>
+        </div>
+    )
+}
+
+function ButtonComponent(props) {
+    return (
+        <button type="button">
+            Set Message
+        </button>
+    )
+}
+```
+
+Let's create two state items in the `WrapperContaner`. 
+
+ - `displayText` as we defined in the earlier example will hold the set message.
+ - `heldText` will keep the inputs typed by the user until they are saved
+
+Pass these and the `setHeldText` setter as `props` to the `DisplayComponent`. 
+
+```
+function WrapperContainer() {
+    // define the displayText to be shown
+    const [ displayText, setDisplayText ] = useState('')
+    // create a holder for unsaved text
+    const [ heldText, setHeldText ] = useState('')
+
+    return (
+        <div>
+            <div className="layout-left">
+                <DisplayComponent 
+                    heldText={heldText} 
+                    setHeldText={setHeldText}
```

Comment:
It may make things clearer to rename `heldText` to `unsavedText` and `displayText` to `savedText`.